HybridCache - minor post-PR fixes

mgravell opened the pull request:
This is supplementary post-PR (it was auto-merged) tweaks requested in https://github.com/dotnet/aspnetcore/pull/55147/, mostly for @amcasey - additional comments, etc

---

Additional notes on 7f8fbc641633693678ded1f2feab355bac7bda8e - this removes any special treatment of `IDisposable`; this is because trying to consider it causes semantic problems; the caller can't know the lifetime, so if we *did* this, we would have a scenario where:

``` c#
var obj = cache.GetOrCreateAsync(...);
// ... other stuff
obj.DoSomething(); // boom ObjectDisposedException
```

because during the "... other stuff", the shared instance got evicted from cache and disposed; this is unpredictable and not a supportable scenario. The correct behaviours are:

1. the caller gets a new instance and disposes per-call, i.e. `using var obj = cache.GetOrCreateAsync(...)`, or
2. it doesn't get disposed at all, using GC and finalizers if necessary

(the latter being less "correct" than the former)

Both of these are what we *already get* from simply not considering `IDisposable` *at all*; 1. is what we get if the type is considered mutable, and 2. is what we get if it is considered immutable; so: it is already in the consumer's hands.

In reality, it **very much** isn't worth worrying to much about `IDisposable` here; the intended use-case of cache is for transient state data, typically POCOs - not connected services etc (we're not the DI layer). IMO we should not overly distort any part of the design worrying about a pathological and hypothetical use-case that is *probably* best avoided with the words: "don't do that".

---

additional notes on 223f1bd8ba158328b0d9c9af35035825fd159b0c - this avoids having to deserialize inside the sync-lock (which is horrible), but to do that we need a single unified ref-count rather than the two separate ref-counts we had previously (one for the stampede state, one for the cache item), so that we can speculatively reserve our value *without* having to deserialize it; this actually *simplifies* the logic, so is desirable in its own right.

mgravell committed:
HybridCache - post-PR fixes

AnthonyLloyd commented:
Hi, I may easily be missing something but don't you have to check the local cache again after you find you are the stampede that needs to do the work? There's a race between the dictionaries where in between someone came in and added the value and removed themselves from the `_currentOperations` before you got there. Or maybe this is very unlikely.

mgravell commented:
@AnthonyLloyd can you outline the scenario you're thinking of, that isn't guarded by the interlocked join and the double-checked locking? The only scenario I can see there is when an in-flight operations is on the cusp of terminating, which means the try-join inside the double-checked lock fails. If that is the scenario you're thinking of, we *could* potentially take the lock for the final decrement+remove. That *is* rare but theoretically non-zero - I could try to measure impact of adding a lock here (probably negligible because we try gard to minimize the locks on the competing path). Or are you thinking of another scenario?

AnthonyLloyd commented:
@mgravell I think it's what you describe:
Thread 1: Has got all the way to just before Cache.SetL1.
Thread 2: Sees nothing in `_localCache` and starts to call `GetOrCreateStampedeState`.
Thread 1: Completes Cache.SetL1 and removes itself from `_currentOperations` before Thread 2 accesses `_currentOperations`.
This would give a completed value in `_localCache` with Thread 2 working to get a duplicate.

mgravell commented:
OK. Probably won't me ready in time for preview 4, but I'll see what the impact of preventing that scenario is.

mgravell commented:
Fun fact: in an incomplete commit earlier, I actually had a partitioned sync-lock grid to avoid any remaining contention on the lock - I ditched that when the only use was the double-check, which is itself rare. If we are going to lock on the remove path, it is probably worth resurrecting that; just a simple partition just using the low 3 bits of the key's hashcode.

mgravell committed:
use partitioned lock to avoid minor race condition on "add" vs "remove" paths

mgravell commented:
@AnthonyLloyd eeaf8571b17348aebd51be46597642e2d2b5bb88

mgravell committed:
tyop

mgravell committed:
1. double-check any use L1 inside the partitioned lock, to allow for the fact that an outgoing previous operation may have added it
2. remove any reference to IDisposable; that creates semantically invalid outcomes

AnthonyLloyd commented:
@mgravell Does it not still have the problem that mutliple threads could come in at the same time, see a local cache miss and then all queue up on the lock and repeat the call? Is the idea that the L2 call can be repeated? And if there is no remote L2 you use L2 = L1 and hence check the local cache again? Sorry I've not studied the code fully.

mgravell commented:
@AnthonyLloyd  the problem there, though, is that I'm not very happy with running "deserialize" inside the sync-lock; I can fix that, but it involves moving a few more pieces; sigh

mgravell committed:
avoid deserialize inside lock, involves refactoring cache item so we can reserve against a cache-item

mgravell committed:
add missing accessibility; move a factory method

mgravell committed:
examples constructed for release notes (to prove they work)

amcasey commented:
CI failure is in `Microsoft.Extensions.Caching.Hybrid.Tests.StampedeTests.MutableTypesNeverShareFinalTask(withCancelation: True)`, which looks related.

amcasey commented:
The new comments look great - thanks!  The new code looks non-trivial and I'll have to look at it more carefully.  @BrennanConroy will probably be interested too (I think you directed his attention here in comments on the original PR).

mgravell commented:
I'll look into the CI in the morning, ta

amcasey commented on src/Caching/Hybrid/src/Internal/DefaultHybridCache.CacheItem.cs:
```diff
@@ -10,37 +11,81 @@ partial class DefaultHybridCache
+            var oldValue = Volatile.Read(ref _refCount);
+            do
+            {
+                if (oldValue is 0 or -1)
```
I think this check is the reason we couldn't just use `Interlocked.Increment`?

mgravell replied:
yes; we never want to increment *from* zero, nor do we ever want to overflow *back* to zero

mgravell replied:
comment added: d8c1d9e11d3edb4ff36d8f30ec56ad5148c5cdc7

mgravell commented on src/Caching/Hybrid/src/Internal/DefaultHybridCache.CacheItem.cs:
```diff
@@ -10,37 +11,81 @@ partial class DefaultHybridCache
+        public bool Release() // returns true ONLY for the final release step
+        {
+            var newCount = Interlocked.Decrement(ref _refCount);
+            if (newCount == 0)
```
fundamentally, the reserve/release semantics; but in the case of getting it wrong, `TryReserve` never pushes us *back* to zero, so: we'd need to screw up 4 billion times; probably warrants a check and error, though

mgravell replied:
Well you weren't wrong: d8c1d9e11d3edb4ff36d8f30ec56ad5148c5cdc7

mgravell committed:
fix comments for @amcasey (and fix a "oops we *were* going negative, my bad")
